### ISSUE TITLE
[Task 15] clarification: make Megatron FP16 optimizer settings configurable

### ISSUE THREAD

**Issue by CCCPIII** [redai-infra/Relax#126]
## 1. Repository, baseline, and branch

- GitHub ID: `CCCPIII`
- Claim: `https://github.com/redai-infra/Relax/issues/86#issuecomment-5077788319`
- Upstream repository: `https://github.com/redai-infra/Relax`
- Working fork: `https://github.com/CCCPIII/Relax`
- Baseline commit: `cfda05351d4b137f60e2b71d048beab1316d9d9f`
- Development branch: `contrib/task15-fp16-config`

## 2. Reproduction environment

- Host OS: Ubuntu 22.04, Linux 6.2.0-37-generic
- Official image:
  `ghcr.io/redai-infra/relaxrl:dev-20260715-8325919e`
- Image digest:
  `sha256:3fa8ce578acda6c829b83016bde42c38fa892681e4f36ca330f545616fe578e2`
- Python: 3.12.3 in the official image
- PyTorch: `2.11.0+cu129`
- CUDA runtime: 12.9.1
- Megatron Core: `0.18.0`
- Transformer Engine: `2.14.1`
- Ray: `2.56.0`
- SGLang: `0.5.12.post1`
- NVIDIA driver: 550.127.05
- Available GPU: 1 × NVIDIA A800-SXM4-40GB, compute capability 8.0
- CPU: 16 logical CPUs
- Memory: 31 GiB

The available machine does not match the target recipe's eight-GPU topology.
The PR will therefore provide reproducible CPU unit-test evidence, shell
validation, and a real parser-to-optimizer configuration test. It will not
claim a single- or eight-GPU end-to-end result.

## 3. Commands

Focused test command:

```bash
python -m pytest -v \
  tests/utils/test_arguments_fp16_optimizer.py \
  tests/backends/megatron/test_fp16_optimizer_config.py \
  tests/scripts/test_run_qwen3_fp16_script.py \
  tests/utils/test_arguments_opd_teacher_colocate.py \
  tests/backends/megatron/test_mtp_sft_forward_kwargs.py \
  tests/backends/megatron/test_mtp_rl_forward_kwargs.py
```

Script validation:

```bash
bash -n scripts/training/text/run-qwen3-4B-fp16-8xgpu.sh
```

Target training entrypoint:

```bash
bash scripts/training/text/run-qwen3-4B-fp16-8xgpu.sh
```

The target entrypoint contains the complete modified optimizer argument set.
It is documented for the required eight-GPU topology but was not executed on
the available one-GPU runner. No end-to-end or performance result will be
claimed from that machine.

The target script explicitly supplies:

```text
--initial-loss-scale 32768
--min-loss-scale 1
--use-precision-aware-optimizer
--no-store-param-remainders
```

Complete user-override example:

```bash
bash scripts/training/text/run-qwen3-4B-fp16-8xgpu.sh \
  --initial-loss-scale 65536 \
  --min-loss-scale 2 \
  --no-use-precision-aware-optimizer \
  --store-param-remainders
```

Changed-file quality checks:

```bash
ruff check \
  relax/utils/arguments.py \
  relax/backends/megatron/model.py \
  tests/utils/test_arguments_fp16_optimizer.py \
  tests/backends/megatron/test_fp16_optimizer_config.py \
  tests/scripts/test_run_qwen3_fp16_script.py
ruff format --check \
  relax/utils/arguments.py \
  relax/backends/megatron/model.py \
  tests/utils/test_arguments_fp16_optimizer.py \
  tests/backends/megatron/test_fp16_optimizer_config.py \
  tests/scripts/test_run_qwen3_fp16_script.py
pre-commit run --all-files
```

## 4. Files, functions, scope, and exclusions

Core functions:

- `relax/utils/arguments.py`
  - `reset_arg()`
  - `get_slime_extra_args_provider()` / nested `add_train_arguments()`
  - `_normalize_precision_optimizer_args()`
  - `slime_validate_args()`
- `relax/backends/megatron/model.py`
  - `_build_optimizer_config_kwargs()`
  - `setup_model_and_optimizer()`

Planned changes:

- `relax/utils/arguments.py`
  - reuse Megatron's native loss-scale and precision-aware options;
  - add the missing negative precision-aware option;
  - add positive/negative `store-param-remainders` options;
  - normalize omitted values before Megatron validation;
  - validate invalid FP16 scale values and scale ordering;
  - emit one combined warning when FP16 compatibility fallback is used.
- `relax/backends/megatron/model.py`
  - reflect fields from the validated argument namespace;
  - stop overwriting optimizer policy and precision values.
- `scripts/training/text/run-qwen3-4B-fp16-8xgpu.sh`
  - explicitly specify `32768 / 1 / enabled / disabled`;
  - forward trailing user arguments so those values remain overridable.
- Focused CPU tests under `tests/utils/` and `tests/backends/megatron/`.
- Script contract tests under `tests/scripts/`, including an executed fake-Ray
  argv capture that verifies trailing overrides and shell word preservation.

Explicit exclusions:

- no CUDA, driver, Megatron, Transformer Engine, or SGLang changes;
- no unrelated optimizer or training behavior;
- no edits to other recipes;
- no performance claim or eight-GPU training claim.

## 5. Baseline, acceptance metrics, and target

Baseline behavior:

- `setup_model_and_optimizer()` unconditionally overwrites
  `initial_loss_scale`, `min_loss_scale`,
  `use_precision_aware_optimizer`, and `store_param_remainders` whenever
  `args.fp16` is true.
- The target Qwen3-4B FP16 script supplies only `--fp16`, so users cannot
  control those four values through that recipe.

Acceptance:

- no optimizer or precision argument overwrite remains in model setup;
- target script explicitly preserves the legacy values and accepts user
  overrides;
- omitted FP16 values use the documented compatibility fallback with one
  readable warning;
- explicit values survive unchanged;
- invalid FP16 scale values and scale ordering fail clearly;
- BF16 receives Megatron-compatible defaults without an FP16 warning;
- focused tests, existing related tests, shell syntax, and required quality
  checks pass.

Measurement and target:

| Requirement | Measurement | Target |
| --- | --- | --- |
| Omitted FP16 configuration | captured fallback warnings | exactly `1` combined warning |
| Explicit FP16 configuration | captured fallback warnings | `0` warnings |
| BF16/FP32 compatibility | captured FP16 fallback warnings | `0` warnings |
| Argument propagation | Namespace vs. `OptimizerConfig` fields | all four fields exactly equal |
| User override | executed target-script argv and parsed config | last supplied value wins |
| Focused/related suite | pytest failures and skips | `0 failed`, `0 skipped` |
| Shell and quality checks | process exit status | exit code `0` |

Tests use deterministic parser/configuration inputs. No random sampling or
performance statistics are used.

`store_param_remainders=True` is forwarded unchanged when explicitly selected.
In the pinned Transformer Engine it is effective only for BF16 parameters with
FP32 master weights and is a documented no-op for FP16 parameters; Relax does
not redefine that upstream configuration as invalid.

This is a correctness/configuration task, not a performance task. Throughput,
latency, GPU utilization, peak-memory, and before/after performance targets are
therefore N/A.

## 6. Delivery

Delivery will be this clarification Issue followed by a Draft Pull Request
containing implementation, focused and related tests, the explicit recipe
configuration, raw test output, compatibility notes, risk, and rollback
instructions. The PR will remain Draft until CI passes and blocking review
comments are resolved.

## Questions for maintainers

1. May omitted FP16 values preserve the existing behavior through the following
   one-warning compatibility fallback: `32768 / 1 / precision-aware enabled /
   parameter remainders disabled`?
2. Is argparse last-value-wins the expected override contract for the target
   recipe, including forwarding an explicit `--store-param-remainders` value
   unchanged even though the pinned Transformer Engine treats it as a no-op for
   FP16 parameters?
3. Should new loss-scale value/order validation remain limited to FP16 so that
   BF16/FP32 retain their existing, unconsumed-value compatibility?

**Comment by CCCPIII** [comment]
Current update for PR #130:

- Head: `d6a2eab50295ddbae237943ed6d331a29e092a8e`
- Tree: `ed3469c21efd7fa993ad9fa2c4ab0d03c7e9d009`
- The precision/default logic has one resolver in `relax.backends.megatron.arguments`. Normal CLI values resolve once in the backend; custom YAML values resolve once after the YAML override.
- The follow-up commit only fixes test isolation. A temporary `relax.core.registry` reload now restores the original module and parent-package attribute, so later tests do not compare Enum classes from different module instances.

Verification:

- [CI 31185164155](https://github.com/redai-infra/Relax/actions/runs/31185164155): all five jobs passed. Python 3.10, 3.11, and 3.12 each reported 858 passed / 275 skipped.
- Exact issue image: root run 1362 passed / 15 skipped / 0 failed; targeted parser/Megatron run 149 passed / 0 skipped.
- Current latest image: root run 1362 passed / 15 skipped / 0 failed; targeted parser/Megatron run 149 passed / 0 skipped.
- The root runs used `pytest .` with no ignore, deselection, or `-x`, on the real RTX 4090 CUDA path.
- All 164 shell files passed `bash -n`; `compileall`, Ruff, `git diff --check`, and Gitleaks checks passed.

Evidence and sanitized JUnit files: https://gist.github.com/CCCPIII/501e05e9ed5316a97c340cef58caca16

The remaining 15 skips are existing image/dependency/model guards and are itemized in the evidence README. Maintainer confirmation is still requested before merge.
